checker/src/main/java/org/checkerframework/checker/fenum/FenumAnnotatedTypeFactory.java

```diff
@@ -77,37 +93,78 @@ public FenumAnnotatedTypeFactory(BaseTypeChecker checker) {
-         * These annotations are declared to be subtypes of FenumTop, using the
-         * @SubtypeOf annotation.
-         * However, there is no way to declare that it is a supertype of Bottom.
-         * Therefore, we use the constructor of GraphQualifierHierarchy that allows
-         * us to set a dedicated bottom qualifier.
+        /** QualifierKind for {@link Fenum} qualifier. */
+        private final QualifierKind FENUM_KIND;
+
+        /**
+         * Creates FenumQualifierHierarchy.
+         *
+         * @param qualifierClasses qualifier classes
+         * @param elements element utils
          */
-        public FenumQualifierHierarchy(MultiGraphFactory factory) {
-            super(factory, FENUM_BOTTOM);
+        public FenumQualifierHierarchy(
+                Collection<Class<? extends Annotation>> qualifierClasses, Elements elements) {
+            super(qualifierClasses, elements);
+            this.FENUM_KIND =
+                    this.qualifierKindHierarchy.getQualifierKind(Fenum.class.getCanonicalName());
         }
 
         @Override
-        public boolean isSubtype(AnnotationMirror subAnno, AnnotationMirror superAnno) {
-            if (AnnotationUtils.areSameByName(superAnno, FENUM)
-                    && AnnotationUtils.areSameByName(subAnno, FENUM)) {
-                return AnnotationUtils.areSame(superAnno, subAnno);
-            }
-            // Ignore annotation values to ensure that annotation is in supertype map.
-            if (AnnotationUtils.areSameByName(superAnno, FENUM)) {
-                superAnno = FENUM;
+        protected QualifierKindHierarchy createQualifierKindHierarchy(
+                @UnderInitialization FenumQualifierHierarchy this,
+                Collection<Class<? extends Annotation>> qualifierClasses) {
+            return new QualifierKindHierarchy(qualifierClasses, FenumBottom.class);
+        }
+
+        @Override
+        protected boolean isSubtype(
+                AnnotationMirror subAnno,
+                QualifierKind subKind,
+                AnnotationMirror superAnno,
+                QualifierKind superKind) {
+            return AnnotationUtils.areSame(subAnno, superAnno);
+        }
+
+        @Override
+        protected AnnotationMirror leastUpperBound(
+                AnnotationMirror a1,
+                QualifierKind qual1,
+                AnnotationMirror a2,
+                QualifierKind qual2) {
+            if (qual1 == FENUM_KIND && qual2 == FENUM_KIND) {
+                if (AnnotationUtils.areSame(a1, a2)) {
+                    return a1;
+                }
+                return FENUM_TOP;
+            } else if (qual1 == FENUM_KIND) {
+                return a1;
+            } else if (qual2 == FENUM_KIND) {
+                return a2;
             }
-            if (AnnotationUtils.areSameByName(subAnno, FENUM)) {
-                subAnno = FENUM;
+            throw new BugInCF("Unexpected QualifierKinds %s %s", qual1, qual2);
```

Comment:
Should these now be `TypeSystemErrors`?

Reply:
Ping.

Reply:
This could be a bug in QualifierKindHierarchy or ComplexQualifierHierarchy, so it should stay a BugInCF.